Importer fails to insert a token due to incorrect default value of token type

### Description

Importer failed to insert a token from a `TokenCreate` transaction because the sql generated somehow set `CONTRACT` as `token.type` default value

```
Aug 10 16:51:45 mirrornode-stable-testnet-failover-parser-1 java[13479]: 2022-08-10T10:51:45.281-0600 ERROR scheduling-2 c.h.m.i.p.r.RecordFileParser Error parsing file 2022-08-10T16_42_40.372679003Z.rcd after 33.19 ms com.hedera.mirror.importer.exception.ParserException: Error copying 1 items to table token_temp
        at com.hedera.mirror.importer.parser.batch.BatchInserter.persist(BatchInserter.java:112)
        at com.hedera.mirror.importer.parser.batch.CompositeBatchPersister.persist(CompositeBatchPersister.java:62)
        at com.hedera.mirror.importer.parser.record.entity.sql.SqlEntityListener.executeBatches(SqlEntityListener.java:291)
        at com.hedera.mirror.importer.parser.record.entity.sql.SqlEntityListener.onEnd(SqlEntityListener.java:209)
        at com.hedera.mirror.importer.parser.record.entity.sql.SqlEntityListener.onEnd(SqlEntityListener.java:88)
        at com.hedera.mirror.importer.parser.record.RecordFileParser.doParse(RecordFileParser.java:142)
        at com.hedera.mirror.importer.parser.record.RecordFileParser.doParse(RecordFileParser.java:52)
        at com.hedera.mirror.importer.parser.AbstractStreamFileParser.parse(AbstractStreamFileParser.java:79)
        at com.hedera.mirror.importer.parser.record.RecordFileParser.parse(RecordFileParser.java:117)
        at com.hedera.mirror.importer.parser.record.RecordFileParser.parse(RecordFileParser.java:52)
        at com.hedera.mirror.importer.parser.record.RecordFileParser$$FastClassBySpringCGLIB$$8b9310db.invoke(<generated>)
        at org.springframework.cglib.proxy.MethodProxy.invoke(MethodProxy.java:218)
        at org.springframework.aop.framework.CglibAopProxy$CglibMethodInvocation.invokeJoinpoint(CglibAopProxy.java:793)
        at org.springframework.aop.framework.ReflectiveMethodInvocation.proceed(ReflectiveMethodInvocation.java:163)
        at org.springframework.aop.framework.CglibAopProxy$CglibMethodInvocation.proceed(CglibAopProxy.java:763)
        at org.springframework.transaction.interceptor.TransactionInterceptor$1.proceedWithInvocation(TransactionInterceptor.java:123)
        at org.springframework.transaction.interceptor.TransactionAspectSupport.invokeWithinTransaction(TransactionAspectSupport.java:388)
        at org.springframework.transaction.interceptor.TransactionInterceptor.invoke(TransactionInterceptor.java:119)
        at org.springframework.aop.framework.ReflectiveMethodInvocation.proceed(ReflectiveMethodInvocation.java:186)
        at org.springframework.aop.framework.CglibAopProxy$CglibMethodInvocation.proceed(CglibAopProxy.java:763)
        at org.springframework.retry.interceptor.RetryOperationsInterceptor$1.doWithRetry(RetryOperationsInterceptor.java:97)
        at org.springframework.retry.support.RetryTemplate.doExecute(RetryTemplate.java:329)
        at org.springframework.retry.support.RetryTemplate.execute(RetryTemplate.java:209)
        at org.springframework.retry.interceptor.RetryOperationsInterceptor.invoke(RetryOperationsInterceptor.java:133)
        at org.springframework.retry.annotation.AnnotationAwareRetryOperationsInterceptor.invoke(AnnotationAwareRetryOperationsInterceptor.java:160)
        at org.springframework.aop.framework.ReflectiveMethodInvocation.proceed(ReflectiveMethodInvocation.java:186)
        at org.springframework.aop.framework.CglibAopProxy$CglibMethodInvocation.proceed(CglibAopProxy.java:763)
        at org.springframework.aop.framework.CglibAopProxy$DynamicAdvisedInterceptor.intercept(CglibAopProxy.java:708)
        at com.hedera.mirror.importer.parser.record.RecordFileParser$$EnhancerBySpringCGLIB$$d334e546.parse(<generated>)
        at com.hedera.mirror.importer.config.MessagingConfiguration.lambda$integrationFlow$1(MessagingConfiguration.java:107)
        at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
        at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:77)
        at java.base/jdk.internal.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
        at java.base/java.lang.reflect.Method.invoke(Method.java:568)
        at org.springframework.integration.handler.LambdaMessageProcessor.processMessage(LambdaMessageProcessor.java:105)
        at org.springframework.integration.handler.ServiceActivatingHandler.handleRequestMessage(ServiceActivatingHandler.java:105)
        at org.springframework.integration.handler.AbstractReplyProducingMessageHandler.handleMessageInternal(AbstractReplyProducingMessageHandler.java:136)
        at org.springframework.integration.handler.AbstractMessageHandler.handleMessage(AbstractMessageHandler.java:56)
        at org.springframework.integration.endpoint.PollingConsumer.handleMessage(PollingConsumer.java:158)
        at org.springframework.integration.endpoint.AbstractPollingEndpoint.messageReceived(AbstractPollingEndpoint.java:475)
        at org.springframework.integration.endpoint.AbstractPollingEndpoint.doPoll(AbstractPollingEndpoint.java:461)
        at org.springframework.integration.endpoint.AbstractPollingEndpoint.pollForMessage(AbstractPollingEndpoint.java:413)
        at org.springframework.integration.endpoint.AbstractPollingEndpoint.lambda$createPoller$4(AbstractPollingEndpoint.java:348)
        at org.springframework.integration.util.ErrorHandlingTaskExecutor.lambda$execute$0(ErrorHandlingTaskExecutor.java:57)
        at org.springframework.core.task.SyncTaskExecutor.execute(SyncTaskExecutor.java:50)
        at org.springframework.integration.util.ErrorHandlingTaskExecutor.execute(ErrorHandlingTaskExecutor.java:55)
        at org.springframework.integration.endpoint.AbstractPollingEndpoint.lambda$createPoller$5(AbstractPollingEndpoint.java:341)
        at org.springframework.scheduling.support.DelegatingErrorHandlingRunnable.run(DelegatingErrorHandlingRunnable.java:54)
        at org.springframework.scheduling.concurrent.ReschedulingRunnable.run(ReschedulingRunnable.java:95)
        at java.base/java.util.concurrent.Executors$RunnableAdapter.call(Executors.java:539)
        at java.base/java.util.concurrent.FutureTask.run(FutureTask.java:264)
        at java.base/java.util.concurrent.ScheduledThreadPoolExecutor$ScheduledFutureTask.run(ScheduledThreadPoolExecutor.java:304)
        at java.base/java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1136)
        at java.base/java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:635)
        at java.base/java.lang.Thread.run(Thread.java:833)
Caused by: com.hedera.mirror.importer.exception.ParserException: Error copying 1 items to table token
        at com.hedera.mirror.importer.parser.batch.BatchUpserter.persistItems(BatchUpserter.java:111)
        at com.hedera.mirror.importer.parser.batch.BatchInserter.persist(BatchInserter.java:108)
        ... 54 more
Caused by: org.postgresql.util.PSQLException: ERROR: invalid input value for enum token_type: "CONTRACT"
  Position: 1881
        at org.postgresql.core.v3.QueryExecutorImpl.receiveErrorResponse(QueryExecutorImpl.java:2675)
        at org.postgresql.core.v3.QueryExecutorImpl.processResults(QueryExecutorImpl.java:2365)
        at org.postgresql.core.v3.QueryExecutorImpl.execute(QueryExecutorImpl.java:355)
        at org.postgresql.jdbc.PgStatement.executeInternal(PgStatement.java:490)
        at org.postgresql.jdbc.PgStatement.execute(PgStatement.java:408)
        at org.postgresql.jdbc.PgPreparedStatement.executeWithFlags(PgPreparedStatement.java:167)
        at org.postgresql.jdbc.PgPreparedStatement.executeUpdate(PgPreparedStatement.java:135)
        at com.zaxxer.hikari.pool.ProxyPreparedStatement.executeUpdate(ProxyPreparedStatement.java:61)
        at com.zaxxer.hikari.pool.HikariProxyPreparedStatement.executeUpdate(HikariProxyPreparedStatement.java)
        at com.hedera.mirror.importer.parser.batch.BatchUpserter.insertToFinalTable(BatchUpserter.java:117)
        at com.hedera.mirror.importer.parser.batch.BatchUpserter.insertItems(BatchUpserter.java:160)
        at com.hedera.mirror.importer.parser.batch.BatchUpserter.persistItems(BatchUpserter.java:103)
        ... 55 more
```

The generated sql to insert token:

```sql
with existing as (
  select e.created_timestamp as e_created_timestamp,e.decimals as e_decimals,e.fee_schedule_key as e_fee_schedule_key,e.freeze_default as e_freeze_default,e.freeze_key as e_freeze_key,e.initial_supply as e_initial_supply,e.kyc_key as e_kyc_key,e.max_supply as e_max_supply,e.modified_timestamp as e_modified_timestamp,e.name as e_name,e.pause_key as e_pause_key,e.pause_status as e_pause_status,e.supply_key as e_supply_key,e.supply_type as e_supply_type,e.symbol as e_symbol,e.token_id as e_token_id,e.total_supply as e_total_supply,e.treasury_account_id as e_treasury_account_id,e.type as e_type,e.wipe_key as e_wipe_key, t.*
  from token_temp t
  left join token e on e.token_id = t.token_id
)
insert into token (created_timestamp,decimals,fee_schedule_key,freeze_default,freeze_key,initial_supply,kyc_key,max_supply,modified_timestamp,name,pause_key,pause_status,supply_key,supply_type,symbol,token_id,total_supply,treasury_account_id,type,wipe_key)
select coalesce(created_timestamp, e_created_timestamp, null),coalesce(decimals, e_decimals, null),coalesce(fee_schedule_key, e_fee_schedule_key, null),coalesce(freeze_default, e_freeze_default, false),coalesce(freeze_key, e_freeze_key, null),coalesce(initial_supply, e_initial_supply, null),coalesce(kyc_key, e_kyc_key, null),coalesce(max_supply, e_max_supply, '9223372036854775807'),coalesce(modified_timestamp, e_modified_timestamp, null),coalesce(name, e_name, null),coalesce(pause_key, e_pause_key, null),coalesce(pause_status, e_pause_status, 'NOT_APPLICABLE'),coalesce(supply_key, e_supply_key, null),coalesce(supply_type, e_supply_type, 'INFINITE'),coalesce(symbol, e_symbol, null),coalesce(token_id, e_token_id, null),case when total_supply >= 0 then total_supply else e_total_supply + coalesce(total_supply, 0) end,coalesce(treasury_account_id, e_treasury_account_id, null),coalesce(type, e_type, 'CONTRACT'),coalesce(wipe_key, e_wipe_key, null)
from existing
where coalesce(created_timestamp, e_created_timestamp) is not null
on conflict (token_id)
do update set fee_schedule_key = excluded.fee_schedule_key,freeze_key = excluded.freeze_key,kyc_key = excluded.kyc_key,modified_timestamp = excluded.modified_timestamp,name = excluded.name,pause_key = excluded.pause_key,pause_status = excluded.pause_status,supply_key = excluded.supply_key,symbol = excluded.symbol,total_supply = excluded.total_supply,treasury_account_id = excluded.treasury_account_id,wipe_key = excluded.wipe_key
```

### Steps to reproduce

need to figure it out

### Additional context

_No response_

### Hedera network

testnet

### Version

v0.61.0

### Operating system

_No response_

--- a/hedera-mirror-importer/src/main/java/com/hedera/mirror/importer/repository/upsert/ColumnMetadata.java
+++ b/hedera-mirror-importer/src/main/java/com/hedera/mirror/importer/repository/upsert/ColumnMetadata.java
@@ -24,6 +24,7 @@ import java.text.MessageFormat;
 import java.util.Comparator;
 import java.util.function.BiConsumer;
 import java.util.function.Function;
+import lombok.ToString;
 import lombok.Value;
 import org.apache.commons.lang3.StringUtils;
 
@@ -33,10 +34,12 @@ import com.hedera.mirror.common.domain.UpsertColumn;
 class ColumnMetadata implements Comparable<ColumnMetadata> {
 
     private final Object defaultValue;
+    @ToString.Exclude
     private final Function<Object, Object> getter;
     private final boolean id;
     private final String name;
     private final boolean nullable;
+    @ToString.Exclude
     private final BiConsumer<Object, Object> setter;
     private final Class<?> type;
     private final boolean updatable;

--- a/hedera-mirror-importer/src/main/java/com/hedera/mirror/importer/repository/upsert/EntityMetadataRegistry.java
+++ b/hedera-mirror-importer/src/main/java/com/hedera/mirror/importer/repository/upsert/EntityMetadataRegistry.java
@@ -40,7 +40,6 @@ import javax.persistence.Column;
 import javax.persistence.Entity;
 import javax.persistence.EntityManager;
 import javax.persistence.Id;
-import javax.persistence.Query;
 import javax.persistence.Table;
 import javax.persistence.metamodel.Attribute;
 import javax.persistence.metamodel.EmbeddableType;
@@ -52,6 +51,7 @@ import lombok.extern.log4j.Log4j2;
 import org.apache.commons.lang3.StringUtils;
 import org.hibernate.metamodel.model.domain.spi.SingularPersistentAttribute;
 import org.springframework.core.annotation.AnnotationUtils;
+import org.springframework.jdbc.core.JdbcOperations;
 
 import com.hedera.mirror.common.domain.UpsertColumn;
 import com.hedera.mirror.common.domain.Upsertable;
@@ -63,6 +63,7 @@ public class EntityMetadataRegistry {
 
     private final EntityManager entityManager;
     private final Map<Class<?>, EntityMetadata> entityMetadata = new ConcurrentHashMap<>();
+    private final JdbcOperations jdbcOperations;
 
     public EntityMetadata lookup(Class<?> domainClass) {
         return entityMetadata.computeIfAbsent(domainClass, this::create);
@@ -128,16 +129,21 @@ public class EntityMetadataRegistry {
      * Looks up column defaults in the information_schema.columns table.
      */
     private Map<String, InformationSchemaColumns> getColumnSchema(String tableName) {
-        String sql = "select column_name, regexp_replace(column_default, '::.*', '') as column_default, " +
-                "is_nullable = 'YES' as nullable from information_schema.columns where table_name = ?";
+        String sql = """
+                select column_name, regexp_replace(column_default, '::.*', '') as column_default,
+                is_nullable = 'YES' as nullable from information_schema.columns where table_name = ?
+                """;
 
-        Query query = entityManager.createNativeQuery(sql, InformationSchemaColumns.class);
-        query.setParameter(1, tableName);
-        var schema = (Map<String, InformationSchemaColumns>) query.getResultList()
-                .stream()
+        var columnSchemas = jdbcOperations.query(sql, (rs, rowNum) -> {
+            var columnSchema = new InformationSchemaColumns();
+            columnSchema.setColumnName(rs.getString(1));
+            columnSchema.setColumnDefault(rs.getString(2));
+            columnSchema.setNullable(rs.getBoolean(3));
+            return columnSchema;
+        }, tableName);
+        var schema = columnSchemas.stream()
                 .collect(Collectors.toMap(InformationSchemaColumns::getColumnName, Function.identity()));
-
-        if (schema == null || schema.isEmpty()) {
+        if (schema.isEmpty()) {
             throw new IllegalStateException("Missing information schema for " + tableName);
         }
 

--- a/hedera-mirror-importer/src/main/java/com/hedera/mirror/importer/repository/upsert/EntityMetadataRegistry.java
+++ b/hedera-mirror-importer/src/main/java/com/hedera/mirror/importer/repository/upsert/EntityMetadataRegistry.java
@@ -37,9 +37,7 @@ import java.util.function.Function;
 import java.util.stream.Collectors;
 import javax.inject.Named;
 import javax.persistence.Column;
-import javax.persistence.Entity;
 import javax.persistence.EntityManager;
-import javax.persistence.Id;
 import javax.persistence.Table;
 import javax.persistence.metamodel.Attribute;
 import javax.persistence.metamodel.EmbeddableType;
@@ -199,9 +197,7 @@ public class EntityMetadataRegistry {
     }
 
     @Data
-    @Entity(name = "columns")
     static class InformationSchemaColumns {
-        @Id
         private String columnName;
         private String columnDefault;
         private boolean nullable;

--- a/hedera-mirror-importer/src/test/java/com/hedera/mirror/importer/repository/upsert/EntityMetadataRegistryTest.java
+++ b/hedera-mirror-importer/src/test/java/com/hedera/mirror/importer/repository/upsert/EntityMetadataRegistryTest.java
@@ -25,10 +25,13 @@ import static org.assertj.core.api.Assertions.assertThatCode;
 import static org.assertj.core.api.Assertions.assertThatThrownBy;
 
 import java.util.stream.Stream;
+import javax.persistence.Id;
+import lombok.Data;
 import lombok.RequiredArgsConstructor;
 import org.assertj.core.api.InstanceOfAssertFactories;
 import org.junit.jupiter.api.Test;
 import org.springframework.beans.factory.annotation.Autowired;
+import org.springframework.data.domain.Persistable;
 import org.springframework.transaction.annotation.Transactional;
 
 import com.hedera.mirror.common.domain.UpsertColumn;
@@ -106,6 +109,11 @@ class EntityMetadataRegistryTest extends IntegrationTest {
     }
 
     @Test
+    void nonExisting() {
+        assertThatThrownBy(() -> registry.lookup(NonExisting.class)).isInstanceOf(IllegalStateException.class);
+    }
+
+    @Test
     void notUpsertable() {
         assertThatThrownBy(() -> registry.lookup(Object.class)).isInstanceOf(UnsupportedOperationException.class);
     }
@@ -134,5 +142,18 @@ class EntityMetadataRegistryTest extends IntegrationTest {
 
     @Upsertable
     private static class NonEntity {
+    }
+
+    @Data
+    @javax.persistence.Entity
+    @Upsertable
+    private static class NonExisting implements Persistable<Integer> {
+        @Id
+        private Integer id;
+
+        @Override
+        public boolean isNew() {
+            return true;
+        }
     }
 }

--- a/hedera-mirror-importer/src/test/java/com/hedera/mirror/importer/repository/upsert/EntityMetadataRegistryTest.java
+++ b/hedera-mirror-importer/src/test/java/com/hedera/mirror/importer/repository/upsert/EntityMetadataRegistryTest.java
@@ -24,10 +24,12 @@ import static org.assertj.core.api.Assertions.assertThat;
 import static org.assertj.core.api.Assertions.assertThatCode;
 import static org.assertj.core.api.Assertions.assertThatThrownBy;
 
+import java.util.stream.Stream;
 import lombok.RequiredArgsConstructor;
 import org.assertj.core.api.InstanceOfAssertFactories;
 import org.junit.jupiter.api.Test;
 import org.springframework.beans.factory.annotation.Autowired;
+import org.springframework.transaction.annotation.Transactional;
 
 import com.hedera.mirror.common.domain.UpsertColumn;
 import com.hedera.mirror.common.domain.Upsertable;
@@ -81,6 +83,26 @@ class EntityMetadataRegistryTest extends IntegrationTest {
                 .satisfies(cm -> assertThat(cm.getGetter().apply(entity)).isEqualTo(entity.getAlias()))
                 .satisfies(cm -> assertThatCode(() -> cm.getSetter().accept(entity, newValue))
                         .satisfies(d -> assertThat(entity.getAlias()).isEqualTo(newValue)));
+    }
+
+    @Test
+    @Transactional
+    void lookupSameColumnNameFromMultipleDomainClasses() {
+        // The test case reproduces the issue in the ticket https://github.com/hashgraph/hedera-mirror-node/issues/4265
+        // With the entity manager cache, if we look up two domain classes metadata in the same session, for columns with
+        // the same name, the defaults of the first domain class will override those of the second. For example,
+        // without the fix, entity.type's default will be "'FUNGIBLE_COMMON'"
+        // given, when
+        var tokenMetadata = registry.lookup(Token.class);
+        var entityMetadata = registry.lookup(Entity.class);
+
+        // then
+        var typeDefaultValues = Stream.of(entityMetadata, tokenMetadata)
+                .flatMap(m -> m.getColumns().stream())
+                .filter(c -> "type".equals(c.getName()))
+                .map(ColumnMetadata::getDefaultValue)
+                .toList();
+        assertThat(typeDefaultValues).containsExactly(null, "'FUNGIBLE_COMMON'");
     }
 
     @Test

